### ISSUE TITLE
Hive partitions are not loaded for single cardinality columns

### ISSUE THREAD

**Issue by dshylic** [dask/fastparquet#697]
**What happened**:

I am trying to load a dataset partitioned on year, month, day, and hour.
Example:

```
mydata/year=2021/month=08/day=10/hour=12/filename.parquet
mydata/year=2021/month=09/day=11/hour=12/filename.parquet
mydata/year=2021/month=10/day=12/hour=12/filename.parquet
...
```

Currently, fastparquet cannot handle this task in all cases, as it only reads Hive style partitions as columns if they have *more than one value*. This prevents me from doing things like pushing down a filter for, in my file list there, the `year` column.

**What you expected to happen**:

I expect fastparquet to resemble other libraries (including PyArrow) insofar as it reads all Hive style partition directory values, not just ones with more than one value.

**Minimal Complete Verifiable Example**:

This example creates a partition Parquet directory using the PyArrow library, though the method of creating it is not relevant. You can also create it using fastparquet one file at a time. When it is time to read it back, PyArrow is able to pull out the `year` partition correctly, but `fastparquet` is not.

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({"year": [2021, 2021, 2021, 2021, 2021, 2021], "month": [1,2,3,4,5,6], "val": [1,6,3,4,4,10]})

In [3]: df
Out[3]:
   year  month  val
0  2021      1    1
1  2021      2    6
2  2021      3    3
3  2021      4    4
4  2021      5    4
5  2021      6   10

In [4]: df.to_parquet("/tmp/test.parquet", engine="pyarrow", partition_cols=["year", "month"])

In [5]: %ls -r /tmp/test.parquet
year=2021/

In [6]: %ls /tmp/test.parquet/year=2021/month=3/
1df197ff2f9749968c1c25003eebfc98.parquet  71145151614f42e1876e73c9c6d597aa.parquet

In [7]: pd.read_parquet("/tmp/test.parquet", engine="pyarrow")
Out[7]:
    val  year month
0     1  2021     1
1     1  2021     1
2     6  2021     2
3     6  2021     2
4     3  2021     3
5     3  2021     3
6     4  2021     4
7     4  2021     4
8     4  2021     5
9     4  2021     5
10   10  2021     6
11   10  2021     6

In [8]: pd.read_parquet("/tmp/test.parquet", engine="fastparquet")
Out[8]:
    val month
0     1     1
1     1     1
2     6     2
3     6     2
4     3     3
5     3     3
6     4     4
7     4     4
8     4     5
9     4     5
10   10     6
11   10     6
```

This means that while I can filter by month, for example:

```python
In [9]: pd.read_parquet("/tmp/test.parquet", engine="fastparquet", filters=[("month", "=", 3)])
Out[9]:
   val month
0    3     3
1    3     3
```

I *cannot filter by year*:

```python
In [10]: pd.read_parquet("/tmp/test.parquet", engine="fastparquet", filters=[("year", "=", 2021)])
---------------------------------------------------------------------------
ValueError                                
...
ValueError: No filter can be applied on nonexistent column(s) {'year'}.
```

**Anything else we need to know?**:

I'm not sure if this is expected behavior. If so, a workaround would be much appreciated, as this prevents fastparquet from being applicable for this partitioning use case.

Debugging it shows that the `file_path` is used in `row_groups`. It seems as though the code is looking for the root in which partitions begin and skipping single directory ones. Testing this led to a workaround, in which the `root` can be explicitly set to give expected behavior:

```python
In [1]: import fastparquet

In [2]: fastparquet.ParquetFile("/tmp/test.parquet").info["partitions"]
Out[2]: ['month']

In [3]: fastparquet.ParquetFile("/tmp/test.parquet", root="/tmp/test.parquet").info["partitions"]
Out[3]: ['year', 'month']

In [4]: fastparquet.ParquetFile("/tmp/test.parquet", root="/tmp/test.parquet").to_pandas()
Out[4]:
    val  year month
0     1  2021     1
1     1  2021     1
2     6  2021     2
3     6  2021     2
4     3  2021     3
5     3  2021     3
6     4  2021     4
7     4  2021     4
8     4  2021     5
9     4  2021     5
10   10  2021     6
11   10  2021     6
```

This is fairly brittle though, as you need to, for example, ensure that the root path is absolute and normalize the parquet path itself. For example, from the /tmp directory:

```python
In [1]: import fastparquet

In [2]: fastparquet.ParquetFile("./test.parquet", root="/tmp/test.parquet").info["partitions"]
---------------------------------------------------------------------------
AssertionError     
...
AssertionError: All paths must begin with the given root
```

It also doesn't appear to integrate cleanly (or possibly?) into Pandas' `read_parquet` function.

**Environment**:

fastparquet 0.7.1, installed via pip
Python 3.8
Happens on MacOS and Linux environments
Installed via pip

Thanks!

**Comment by martindurant** [comment]
Indeed, to make it clearer, unless specified (and in the absence of a _metadata file), the code will take the longest path common to all files as the "root". This is because the partitioning is not necessarily hive, but we also have supported "drill"-style, which does not have "key=value" directory names. Maybe we can forget about the past and drill by now.

In any case, when a directory is given, as opposed to a list of files, then that should be the root, and this should be an easy enough fix. Likewise, normalising the input paths should be automatic. It is slightly problematic, because fastparquet supports remote paths but predates fsspec.

I don't know quite what pandas expects when passing kwargs. I suspect it only passes them to the `to_pandas` method.

Would you like implement the first fix, probably setting the root (if None) before [this line](https://github.com/dask/fastparquet/blob/main/fastparquet/api.py#L149)?

**Comment by yohplala** [comment]
> If so, a workaround would be much appreciated, as this prevents fastparquet from being applicable for this partitioning use case.

Also, as a workaround, you could use fastparquet to write as you use it to read.
But you are right that fastparquet should be able to find its way through pyarrow hive-partitionned data.

```python
df.to_parquet(fn, engine="fastparquet", partition_cols=["year", "month"])
pd.read_parquet(fn, engine="fastparquet")
```

Bests
